File: goodAlgorithms/fastPow.py
def fast_power(a, n, m): # (a ^ n) % m
    """
    This function assumes that n >= 0
    """
    result = 1
    value = a
    power = n
    while power > 0:
        if power % 2 == 1:
            result = result * value
            result %= m
        value = value * value
        value %= m
        power = power//2
    return result

File: goodAlgorithms/fastPow.c
long long fast_pow(int a,int n)
{
    long long result = 1;
    long long power = n;
    long long value = a;
    while(power>0)
    {
        if(power&1)
            {result = result*value;
            result = result%1000000007;}
        value = value*value;
        value = value%1000000007;
        power /= 2;
        //power >>= 1;
    }
    return result;
}
